### PR TITLE
fix(Stamp): Gas golfing

### PR DESCRIPTION
**Pull request by n00b21337**
[improv.txt](https://github.com/ethersphere/storage-incentives/files/12428614/improv.txt)
[gas-report.txt](https://github.com/ethersphere/storage-incentives/files/12428616/gas-report.txt)
Optimization of PostageStamp contract for less gas usage.

Added improv diff that shows improvements on most counts with Stamp and final gas report.

### CODE REVIEW

**Review by mfw78 — COMMENTED** [review]
Great gas golfing, but approach with caution: https://github.com/NethermindEth/nethermind/issues/6024

**Comment by n00b21337** [comment]
> Great gas golfing, but approach with caution: [NethermindEth/nethermind#6024](https://github.com/NethermindEth/nethermind/issues/6024)

thnx on this info, good to know. hopefully they resolve this fully.

File: src/PostageStamp.sol [hunk 1]
```diff
@@ -412,22 +421,31 @@ contract PostageStamp is AccessControl, Pausable {
...
+            }
             validChunkCount -= batchSize;
             // since the batch expired _during_ the period we must add
             // remaining normalised payout for this batch only
             pot += batchSize * (batch.normalisedBalance - leb);
             tree.remove(fbi, batch.normalisedBalance);
             delete batches[fbi];
+
+            unchecked {
+                ++i;
```
Comment (mrekucci) [inline src/PostageStamp.sol:438]: Why the increment was moved here and is labeled as unchecked?

File: src/PostageStamp.sol [hunk 2]
```diff
@@ -138,32 +177,34 @@ contract PostageStamp is AccessControl, Pausable {
...
-        require(ERC20(bzzToken).transferFrom(msg.sender, address(this), totalAmount), "failed transfer");
+        if (!ERC20(bzzToken).transferFrom(msg.sender, address(this), totalAmount)) {
+            revert TransferFailed();
+        }
 
-        // normalisedBalance is an absolute value per chunk, as if the batch had existed
-        // since the block the contract was deployed, so we must supplement this batch's
-        // _initialBalancePerChunk with the currentTotalOutPayment()
         uint256 normalisedBalance = currentTotalOutPayment() + (_initialBalancePerChunk);
+        if (normalisedBalance <= 0) {
```
Comment (nugaon) [inline src/PostageStamp.sol:203]: can it be even below zero ever? since it is an uint.

File: src/PostageStamp.sol [hunk 3]
```diff
@@ -222,8 +275,6 @@ contract PostageStamp is AccessControl, Pausable {
             lastUpdatedBlockNumber: block.number
         });
 
-        require(normalisedBalance > 0, "normalisedBalance cannot be zero");
```
Comment (nugaon) [inline src/PostageStamp.sol:225]: you removed from here, but you placed this check in `createBatch`. Why?

**Review by nugaon — COMMENTED** [review]
Overall, great refactor : )

File: src/PostageStamp.sol (hunk 2 above)
Comment (n00b21337) [reply src/PostageStamp.sol:203]: yes u r right.  Just did it auto opposite of >  but it can be just == 
Fixed

File: src/PostageStamp.sol (hunk 3 above)
Comment (n00b21337) [reply src/PostageStamp.sol:225]: Its here same as with copybatch you just didnt see it :)

image

File: src/PostageStamp.sol [hunk 4]
```diff
@@ -59,30 +48,43 @@ contract PostageStamp is AccessControl, Pausable {
...
     uint256 public pot;
 
-    // blocks in 24 hours ~ 24 * 60 * 60 / 5 = 17280
-    uint256 public minimumValidityBlocks = 17280;
+    // Normalised balance at the blockheight expire() was last called.
+    uint256 public lastExpiryBalance;
 
     // Price from the last update.
-    uint256 public lastPrice = 0;
+    uint64 public lastPrice;
```
Comment (nugaon) [inline src/PostageStamp.sol:64]: I wouldn't narrow this variable that much, who knows!

File: src/PostageStamp.sol [hunk 5]
```diff
@@ -313,24 +371,26 @@ contract PostageStamp is AccessControl, Pausable {
...
 
-        lastPrice = _price;
-        lastUpdatedBlock = block.number;
+        lastPrice = uint64(_price);
+        lastUpdatedBlock = uint64(block.number);
 
         emit PriceUpdate(_price);
     }
 
     function setMinimumValidityBlocks(uint256 _value) external {
```
Comment (nugaon) [inline src/PostageStamp.sol:388]: why not change the parameter as well?

**Review by nugaon — APPROVED** [review]
LGTM

**Comment by n00b21337** [comment]
Continued here 
https://github.com/ethersphere/storage-incentives/pull/187

**Review by zelig — APPROVED** [review]

**Comment by n00b21337** [comment]
was overriden in some merge, changes are not in master, will need to reapply

**Pull request merged**